### PR TITLE
Issue 15757 - D main is a nested function and cannot be accessed

### PR DESCRIPTION
**Pull request by 9rnsr**

### CODE REVIEW

**Comment by WalterBright** [comment]
Refactorings, especially 250 line ones, need to go in their own PRs.

**Comment by 9rnsr** [comment]
https://github.com/D-Programming-Language/dmd/pull/5582

@WalterBright You should learn how to ignore indentations in github diff view.

File: src/s2ir.c [hunk 1]
```diff
@@ -44,7 +44,7 @@ elem *exp2_copytotemp(elem *e);
 elem *incUsageElem(IRState *irs, Loc loc);
 elem *addressElem(elem *e, Type *t, bool alwaysCopy = false);
 type *Type_toCtype(Type *t);
-elem *toElemStructLit(StructLiteralExp *sle, IRState *irs, Symbol *sym, bool fillHoles);
+elem *toElemStructLit(StructLiteralExp *sle, IRState *irs, TOK op, Symbol *sym, bool fillHoles);
```
Comment (WalterBright) [inline src/s2ir.c]: not used in s2ir.c, delete it

File: src/e2ir.c [hunk 2]
```diff
@@ -2821,10 +2821,10 @@ elem *toElem(Expression *e, IRState *irs)
                     ex = e1->E1;
                 if (ae->e2->op == TOKstructliteral &&
                     ex->Eoper == OPvar && ex->EV.sp.Voffset == 0 &&
-                    ae->op == TOKconstruct)
+                    ae->op != TOKassign)
```
Comment (WalterBright) [inline src/e2ir.c]: `TOKconstruct` or `TOKblit` is much clearer

**Comment by WalterBright** [comment]
> You should learn how to ignore indentations in github diff view.

Perhaps, but the fact that @yebblies found a bug in #5582 should make it clear why these things should be in separate PRs.

File: src/s2ir.c (hunk 1 above)
Comment (9rnsr) [inline src/s2ir.c]: Done.

File: src/e2ir.c (hunk 2 above)
Comment (9rnsr) [inline src/e2ir.c]: Done.

**Comment by WalterBright** [comment]
Auto-merge toggled on

**Comment by 9rnsr** [comment]
Thanks!

**Pull request merged**